### ISSUE TITLE
"assert sys.platform == 'linux'" incorrectly marks code unreachable

### ISSUE THREAD

**Issue by letmaik** [microsoft/pyright#621]
**Describe the bug**
I'm on Linux and the assert succeeds but pyright believes it's False and hence considers everything after the assert as unreachable code.

**To Reproduce**
See code below.

**Expected behavior**
It should evaluate the assert correctly. I would also like an option that disables unreachability analysis based on such dynamic asserts since this will be different depending on where it runs.

**Screenshots or Code**

```
def foo():
    assert sys.platform == 'linux'
    print("foo") # <- shaded out
```

**VS Code extension or command-line**
 VS Code

**Comment by erictraut** [comment]
Pyright needs to make assumptions about the value of sys.platform during analysis because many type stub files (including the typeshed stub files for stdlib) use sys.platform to conditionalize code for different platforms. By default, Pyright assumes that sys.platform is "" (empty string). This is an appropriate assumption for code that is intended to run on all platforms. If you intend to run your code on a specific platform, you can specify the target in the [pythonconfig.json file using the `pythonPlatform` setting](https://github.com/microsoft/pyright/blob/master/docs/configuration.md). If you set this to `"Linux"`, Pyright will assume that `sys.platform == 'linux'`.

Disabling unreachability analysis entirely isn't an option because there are many pieces of code, including stub files, that rely on unreachability. Disabling unreachability analysis would result in many type analysis errors.

Recommended workarounds:
1. If your code is intended to run only on Linux, then use the pythonPlatform setting on pyrightconfig.json.
2. If parts of your code (submodules) are intended to run only on Linux and other submodules are intended to run on other platforms, define multiple "execution environments" in pyrightconfig.json and specify a different pythonPlatform for each.
3. If your code is intended to run on multiple platforms, don't use an assert. Instead, use conditional logic where necessary to control platform-specific behavior.
4. Avoid expressions that directly compare sys.platform to a string literal. These are special-cased by Pyright and other type checkers. For example, use the following:

```
TARGET_PLATFORM = 'linux'

def foo():
    assert sys.platform == TARGET_PLATFORM
    print("foo") # <- shaded out
```

**Comment by letmaik** [comment]
Thanks a lot for the explanations. That definitely helps. I settled for option 1 for now.

**Comment by WhyNotHugo** [comment]
I'm having a similar issue with a similar scenario.  
The project I'm working on also support win32, and has some code paths like this:

```python
import sys

if sys.platform == "win32":
    print("hi!")
```

I can't move the `"win32"` string into a variable because that would never be approved; it reduces readability just for the sake of a tool I'm using.

Maybe things like `sys.platform` and `sys.version` can be special cased somehow?

**Comment by erictraut** [comment]
Let me make sure I understand what problem you're trying to address. Are you concerned that Pyright is marking these platform-specific code paths as unreachable (displaying them as "grayed out" in the editor)?

The problem is that Pyright needs to make an assumption about what value `sys.platform` will evaluate to at runtime. The standard typeshed stubs are written this way.

I think the best option is for you to pick your most important platform (the one that you special-case the most) and set the pythonPlatform setting on pyrightconfig.json to match that value.
